File: Java/src/DeleteNodesAndReturnForest.java
import java.util.ArrayList;
import java.util.List;

public class DeleteNodesAndReturnForest {
    private List<TreeNode> res;
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        res = new ArrayList<>();
        if(root == null) return res;
        boolean[] delete = new boolean[1001];
        for(int i : to_delete)
            delete[i] = true;
        
        if(!delete[root.val]) res.add(root);
        helper(root, delete);
        
        return res;
    }

    private void helper(TreeNode n, boolean[] delete){
        if(n == null) return;

        TreeNode l = n.left;
        TreeNode r = n.right;
        if(l != null){
            if(delete[l.val])
                n.left = null;
            else if(delete[n.val])
                res.add(l);
            helper(l, delete);
        }
        
        if(r != null){
            if(delete[r.val])
                n.right = null;
            else if(delete[n.val])
                res.add(r);
            helper(r, delete);
        }

    }
    
}

File: Python/DeleteNodesAndReturnForest.py

class Solution:
    def delNodes(self, root: TreeNode, to_delete: List[int]) -> List[TreeNode]:
        self.res = []
        self.d = [False] * 1001

        for i in to_delete:
            self.d[i] = True

        if not self.d[root.val]:
            self.res.append(root)

        self.dfs(root)
        return self.res

    def dfs(self, n: TreeNode):
        
        l = n.left
        r = n.right

        if l != None:
            if not self.d[l.val] and self.d[n.val]:
                self.res.append(l)
            if self.d[l.val]:
                n.left = None
            self.dfs(l)
        
        if r != None:
            if not self.d[r.val] and self.d[n.val]:
                self.res.append(r)
            if self.d[r.val]:
                n.right = None
            self.dfs(r)
            



